### ISSUE TITLE
Planning

### ISSUE THREAD

**Issue by rejasupotaro** [rejasupotaro/amazon-product-search#1]
[AIcrowd | ESCI Challenge for Improving Product Search | Challenges](https://www.aicrowd.com/challenges/esci-challenge-for-improving-product-search)

# Schedule

- 03/25: Dataset Releases
- 04/05: Baselines published
- 05/20: Code submissions
- 06/15: Competition Ends

# TODOs

## Build Pipeline

- [x] Create a repository
- [ ] ~~BigQuery?~~
- [x] Build a pipeline with a random model
- [x] Vertex AI for training
  - https://cloud.google.com/vertex-ai/docs/training/custom-training
  - https://cam-inc.co.jp/p/techblog/582427719648674619
  - https://github.com/daigo0927/blog/tree/191bd8446b4e9591e90be2ee8b1f44ef1db70779/vertex-pipelines-sample
- [x] Colab: https://colab.research.google.com/
- [x] Submission

## Dataset

- Remove noisy data
- Feature scaling
- Main features
  - Query-product graph
  - Language
- Text
  - Clip different sections
  - Max pooling
- Labels
  - Get data from other tasks

## Modeling

- GBDTs?
  - [Are Neural Rankers still Outperformed by Gradient Boosted Decision Trees?](https://research.google/pubs/pub50030/)
  - [tfdf.keras.GradientBoostedTreesModel](https://www.tensorflow.org/decision_forests/api_docs/python/tfdf/keras/GradientBoostedTreesModel)
- TfidfVectorizer
- HashingVectorizer
- Language models
  - Sentence Transformer
    - Cross Encoder
    - https://github.com/amazon-research/esci-code
  - USE
    - https://tfhub.dev/google/universal-sentence-encoder-multilingual/3
    - https://tfhub.dev/google/universal-sentence-encoder/4
  - BERT
    - https://mccormickml.com/2019/05/14/BERT-word-embeddings-tutorial/
    - Multilingual Models: https://tech.yellowback.net/posts/transformers-japanese-models
    - bert-base-multilingual-case
    - distilbert-base-multilingual-cased
    - xlm-roberta-base
    - distilbert-base-multilingual-cased
    - facebook/mbart-large-cc25
    - google/mt5-base
    - facebook/m2m100_418M
    - google/bt5-base
    - `[CLS] question_title [SEP] question_body [SEP] ansewr [SEP]`
    - Use more CLS tokens
      - https://www.kaggle.com/c/google-quest-challenge/discussion/123770
  - XLM
  - GPT
    - DistilGPT2
  - Funnel Transformer large
  - Google Electra
  - LaBSE
    - https://tech.yellowback.net/posts/sentence-transformers-japanese-models
  - https://tech.yellowback.net/posts/sentence-transformers-japanese-models
  - https://tech.yellowback.net/posts/transformers-japanese-models
- Custom classification layer
  - https://github.com/huggingface/transformers/issues/1001
- Fine-tuning
- MatchPyramid
- Main architecture
  - Retrieval => Ranking?
  - FM-based
  - Attention
  - TF-Kernel
  - Max Pooling
  - Weighted Sum
  - [Row skipping + Slower decay](https://www.kdd.org/kdd2020/accepted-papers/view/debiasing-grid-based-product-search-in-e-commerce)
  - [PARADE: Passage Representation Aggregation for Document Reranking](https://arxiv.org/abs/2008.09093)
  - Debiasing (No position bias in labels?)
    - [Handling Position Bias for Unbiased Learning to Rank in Hotels Search](https://arxiv.org/abs/2002.12528)
- ~~Dropout~~
- Loss
  - Point-wise
    - Binary Cross-entropy
    - Categorical Cross-entropy
  - Pair-wise
    - Hinge
  - List-wise
    - Approx NDCG
   - [An Alternative Cross Entropy Loss for Learning-to-Rank](https://arxiv.org/abs/1911.09798)
   - [LambdaLoss](https://dl.acm.org/doi/pdf/10.1145/3269206.3271784?casa_token=5GLC7MH5enoAAAAA:CNy69p1xM4iqAbs_FS7fQbqGJRBPsXlHhgJMQqDpIQ2izpFVx1KOtFVvIx8AWk_h-Hcs1lNNBCI)
- Learning Rate
  - ~~Decay / Scheduling~~: https://keras.io/api/optimizers/
- Metrics
- Pseudo labeling
  - https://github.com/UKPLab/sentence-transformers/blob/cdd1517c61dec93584906e2bc822378e05809eb4/examples/domain_adaptation/README.md
- Ensembling
  - https://www.kaggle.com/code/takamichitoda/26th-place-solution/notebook#Ensemble 

## Training

- Local
- On cloud

## Result Analysis

- Metrics by brand

## Other Resources

- [Improving relevance quality in product search using high-precision query-product semantic similarity - Amazon Science](https://www.amazon.science/publications/improving-relevance-quality-in-product-search-using-high-precision-query-product-semantic-similarity)
- [深層学習による自然言語処理 輪読会#5 資料 - Speaker Deck](https://speakerdeck.com/tok41/shen-ceng-xue-xi-niyoruzi-ran-yan-yu-chu-li-lun-du-hui-number-5-zi-liao)
- Kaggle
  - https://zenn.dev/ycarbon/articles/6b26cc19ff33f5
  - https://www.kaggle.com/code/christofhenkel/how-to-preprocessing-when-using-embeddings
  - https://www.kaggle.com/competitions/quora-insincere-questions-classification/code?competitionId=10737&sortBy=voteCount
  - NBME - Score Clinical Patient Notes
  - Jigsaw Toxic Comment Classification
    - https://engineering.linecorp.com/ja/blog/challenge-the-jigsaw-rate-severity-of-toxic-comments/
  - Jigsaw Unintented Bias in Toxicity Classification
  - Jigsaw MultiLingual Toxic Comment Classification
  - Malignant Comment Classification
  - Real Toxicity Prompts
  - Google QUEST Q&A Labeling
    - https://www.kaggle.com/c/google-quest-challenge/discussion/130076
